json_codec.go

```diff
@@ -0,0 +1,846 @@
+		return st.buildString(enclosingNamespace, t.(string), schema)
+	case map[string]interface{}, []interface{}:
+		// EXAMPLE: "type":{"type":fixed","name":"fixed_16","size":16}
+		// EXAMPLE: "type":["null","int"]
+		return st.buildCodec(enclosingNamespace, t)
+	default:
+		return nil, newCodecBuildError("map", "type ought to be either string, map[string]interface{}, or []interface{}; received: %T", t)
+	}
+}
+
+func (st symtabJSON) buildString(enclosingNamespace, typeName string, schema interface{}) (*codec, error) {
+	switch typeName {
+	case "null":
+		return st.nullCodec, nil
+	case "boolean":
+		return st.booleanCodec, nil
+	case "int":
+		return st.intCodec, nil
+	case "long":
+		return st.longCodec, nil
+	case "float":
+		return st.floatCodec, nil
+	case "double":
+		return st.doubleCodec, nil
+	case "bytes":
+		return st.bytesCodec, nil
+	case "string":
+		return st.stringCodec, nil
+	case "record":
+		return st.makeRecordCodec(enclosingNamespace, schema)
+	case "enum":
+		return st.makeEnumCodec(enclosingNamespace, schema)
+	case "fixed":
+		return st.makeFixedCodec(enclosingNamespace, schema)
+	case "map":
+		return st.makeMapCodec(enclosingNamespace, schema)
+	case "array":
+		return st.makeArrayCodec(enclosingNamespace, schema)
+	default:
+		t, err := newName(nameName(typeName), nameEnclosingNamespace(enclosingNamespace))
+		if err != nil {
+			return nil, newCodecBuildError(typeName, "could not normalize name: %q: %q: %s", enclosingNamespace, typeName, err)
+		}
+		c, ok := st.name[t.n]
+		if !ok {
+			return nil, newCodecBuildError("unknown", "unknown type name: %s", t.n)
+		}
+		return c, nil
+	}
+}
+
+type unionJSONEncoder struct {
+	ef  encoderFunction
+	utn string
+}
+
+// Given a union schema figure out the union type name.
+func getUnionTypeName(friendlyName string, enclosingNamespace string, schema interface{}) (string, error) {
+	// A schema can be a primitive or a complex type.
+	// Primitive types can be encoded as just "primitive" or {"type": "primitive"}.
+
+	switch schema.(type) {
+	case string:
+		// "primitive"
+		return schema.(string), nil
+	}
+
+	schemaJSONMap, ok := schema.(map[string]interface{})
+	if !ok {
+		return "", fmt.Errorf("Unsupported type in union schema %v", schema)
+	}
+
+	unionType, ok := schemaJSONMap["type"]
+	if !ok {
+		return "", fmt.Errorf("Type field is missing in union schema %v", schema)
+	}
+
+	unionTypeName, ok := unionType.(string)
+	if !ok {
+		return "", fmt.Errorf("Type attribute has to be a string in union schema %v", schema)
+	}
+
+	switch unionTypeName {
+	case "record", "enum", "fixed":
+		// The union type name is the fully qualified name of the named type.
+		name, err := newName(nameSchema(schemaJSONMap), nameEnclosingNamespace(enclosingNamespace))
+		if err != nil {
+			return "", err
+		}
+		unionTypeName = name.n
+	}
+
+	return unionTypeName, nil
+}
+
+func (st symtabJSON) makeUnionCodec(enclosingNamespace string, schema interface{}) (*codec, error) {
```

Comment:
Here we go, I see this needs some different behavior. But I think we could still make it work and be a bit less boilerplate by leveraging embedding.